### PR TITLE
Partially revert PR #8809

### PR DESCRIPTION
**Pull request by prestonvanloon**
**What type of PR is this?**

Other

**What does this PR do? Why is it needed?**

Revert production code from #8809, but keep the tests.

**Which issues(s) does this PR fix?**

Part of #8298

**Other notes for review**

#8809 is being reverted as the code did not change the behavior of the deposit lookup. The original code used the standard library to perform a binary search lookup of the desired deposit index while the new code uses a linear search starting from the back of the list. 

While #8809 could be an improvement, especially if this list is long and the found elements are typically at the very end of the list, we do not have the supporting benchmarks or data to support the changes at this time. 

Given the sensitivity of the incident, I am proposing we revert the changes introduced in #8809 for now.

### CODE REVIEW

**Review by terencechain — APPROVED** [review]

**Comment by prestonvanloon** [comment]
Please do not merge until all assigned team members approve

**Review by farazdagi — APPROVED** [review]

**Review by nisdas — APPROVED** [review]

**Review by nisdas — APPROVED** [review]

File: beacon-chain/cache/depositcache/deposits_cache.go
Comment (prestonvanloon) [reply beacon-chain/cache/depositcache/deposits_cache.go:177]: I don't think this is the case.

https://github.com/prysmaticlabs/prysm/compare/pr-feedback-8829

https://play.golang.org/p/8Tnwk_oDK5t

File: beacon-chain/cache/depositcache/deposits_cache.go
Comment (prestonvanloon) [reply beacon-chain/cache/depositcache/deposits_cache.go:180]: Please see other feedback, it seems that 0 index is returned with nothing is found. https://play.golang.org/p/8Tnwk_oDK5t

Comment (rkapka) [reply beacon-chain/cache/depositcache/deposits_cache.go:180]: > Search uses binary search to find and return the smallest index i in [0, n) at which f(i) is true, assuming that on the range [0, n), f(i) == true implies f(i+1) == true. That is, Search requires that f is false for some (possibly empty) prefix of the input range [0, n) and then true for the (possibly empty) remainder; Search returns the first true index. If there is no such index, Search returns n.

In my opinion 0 is indeed a valid index @prestonvanloon

**Pull request merged**
